contracts/javascore/cluster-connection/src/main/java/xcall/adapter/cluster/ClusterConnection.java

```diff
@@ -0,0 +1,359 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xcall.adapter.cluster;
+
+import java.math.BigInteger;
+import score.Context;
+
+import score.Address;
+import score.BranchDB;
+import score.ByteArrayObjectWriter;
+import score.DictDB;
+import score.VarDB;
+import score.ArrayDB;
+import scorex.util.ArrayList;
+
+import score.annotation.EventLog;
+import score.annotation.External;
+import score.annotation.Payable;
+import java.util.List;
+
+
+
+public class ClusterConnection {
+    protected final VarDB<Address> xCall = Context.newVarDB("callService", Address.class);
+    protected final VarDB<Address> adminAddress = Context.newVarDB("admin", Address.class);
+    protected final VarDB<Address> relayerAddress = Context.newVarDB("relayer", Address.class);
+    protected final VarDB<BigInteger> validatorsThreshold = Context.newVarDB("reqValidatorCnt", BigInteger.class);
+    private final VarDB<BigInteger> connSn = Context.newVarDB("connSn", BigInteger.class);
+    private final ArrayDB<Address> validators =  Context.newArrayDB("signers", Address.class);
+
+    protected final DictDB<String, BigInteger> messageFees = Context.newDictDB("messageFees", BigInteger.class);
+    protected final DictDB<String, BigInteger> responseFees = Context.newDictDB("responseFees", BigInteger.class);
+    protected final BranchDB<String, DictDB<BigInteger, Boolean>> receipts = Context.newBranchDB("receipts",
+            Boolean.class);
+    public ClusterConnection(Address _relayer, Address _xCall) {
+        if (xCall.get() == null) {
+            xCall.set(_xCall);
+            adminAddress.set(Context.getCaller());
+            relayerAddress.set(_relayer);
+            connSn.set(BigInteger.ZERO);
+        }
+    }
+
+      /**
+     * Retrieves the validators.
+     *
+     * @return The validators .
+     */
+    @External(readonly = true)
+    public Address[] listValidators() {
+        Address[] sgs = new Address[validators.size()];
+        for(int i = 0; i < validators.size(); i++) {
+            sgs[i] = validators.get(i);
+        }
+        return sgs;
+    }
+
+/**
+ * Adds a list of validators and sets the validation threshold.
+ *
+ * Clears existing validators and adds the provided addresses as validators.
+ * Ensures that the caller is an admin and that the number of validators
+ * meets or exceeds the specified threshold.
+ *
+ * @param _validators an array of addresses to be added as validators
+ * @param _threshold the minimum required number of validators
+ * @throws Exception if the number of validators is less than the threshold
+ */
+    @External
+    public void addValidator(Address[] _validators, BigInteger _threshold) {
+        OnlyAdmin();
+        clearValidators();
+        for (Address validator : _validators) {
+            if(!isValidator(validator)) {
+                validators.add(validator);
+            }
```

Comment:
do we need to check after we have already called clear validators? cant we just insert updated validator list from param?

Reply:
we could have repeated validators in the sent params as well, so need to check!